### ISSUE TITLE
ResourceType vs. 'ProductModel'

### ISSUE THREAD

**Issue by elf-pavlik** [valueflows/resource#26]
https://github.com/valueflows/resource/issues/25#issuecomment-220683038
> We can NOT expect Resource Type to mandate units. For example:

> Resource Type: http://www.productontology.org/doc/Apple.html

> If we think about Input/Output and Offer/Request
* amount: 5kg , type: prodont:Apple
* amount: 24pcs, type:prodont:Apple
* amount: 11lb, type:prodont:Apple

> I agree that agents might want to define default units, which **they prefer to use** for specifying amounts for resources of particular type. Still we can NOT rely on type definition to provide unit which will apply to all instances of that type.

We need to start documenting differences between generic type of a resource and 'product models'. For example generic type can't mandate unit, but more agent specific model will usually have unit.

I propose to include [Olive oil request](https://github.com/valueflows/valueflows/blob/master/use-cases/olive-oil-request.md) use case among those which will drive our exploration.

I request generic type of resource: http://www.productontology.org/doc/Olive_oil.html
3 providers offer particular models which match this generic type

Possibly a rule of thumb: "generic resource types don't relate to any specific agent"

Yesterday on chat @bhaugen confirmed that NRP instances currently don't share definitions of resource types. It suggests me that currently 'ResoruceType' might have more of a 'ProductModel' notion in NRP (speculating heavily here)

**Comment by fosterlynn** [comment]
>It suggests me that currently 'ResoruceType' might have more of a 'ProductModel' notion in NRP (speculating heavily here)

Yes, true.  Although ResourceType includes more than ProductModel, since it includes types of work, and things not for sale.  Not oriented around final product for sale, oriented around contributory accounting and coordination of work.

**Comment by bhaugen** [comment]
> Yesterday on chat @bhaugen confirmed that NRP instances currently don't share definitions of resource types. It suggests me that currently 'ResoruceType' might have more of a 'ProductModel' notion in NRP (speculating heavily here)

Not sharing definitions between instances of NRP is only about technical software limitations and user-community requirements. We want to share a lot of definitions and also have resource flows that spanned instances, but have not had two instances of NRP that wanted to do that.

It's more about the limitations of a relational database, which is what NRP is built on, and not dealing with domains like food yet in NRP that might require a lot of unit conversions.

**Comment by bhaugen** [comment]
In other words, I don't care if Resource Types specify units if some situation does not find them useful.

**Comment by elf-pavlik** [comment]
Do we agree that each instance of Resource needs to specify unit which it uses? We can NOT assume that type definition provides unit which applies to that particular instance.

**Comment by bhaugen** [comment]
> Do we agree that each instance of Resource needs to specify which unit it uses? 

I think so, but need to think back to the times I implemented unit conversions. If I remember correctly, in each case the Resource had a base unit. For food, in particular, people might specify several different units for the same resource, and get their answer converted from the quantity of the base unit into the unit they wanted to see. For example, converting pounds to kilograms or vice versa.

This is a fairly hairy problem, because besides the "standard" conversions, we'll hit currency and product packaging conversions (how many cabbages in a crate, for example).

And I think we will need to do all of that. I don't know much we need to put into the VF vocab and how much to re-use. Certainly the standard conversions can come from qudt.

**Comment by fosterlynn** [comment]
>Do we agree that each instance of Resource needs to specify which unit it uses?

I'm good with this.

**Comment by ahdinosaur** [comment]
[here's an example i just wrote of a "ProductModel"](https://github.com/valueflows/valueflows/pull/115#issuecomment-220808238), am i missing something about why we need a new term for this?

**Comment by elf-pavlik** [comment]
If I cast intent, I would use some common reference for apple eg. http://www.productontology.org/doc/Apple.html

If each farmer defines it's own resource type for apples and doesn't relate it in clear way with a commonly known type definition of apple. How will I get matches for all the apples offered in my region?

You can find more detailed example in [Olive oil request](https://github.com/valueflows/valueflows/blob/master/use-cases/olive-oil-request.md) use case.

**Comment by bhaugen** [comment]
@ahdinosaur 
> am i missing something about why we need a new term for this?

Than what? Product Model or Resource Type (or Apple)?

**Comment by ahdinosaur** [comment]
@bhaugen nevermind, i wasn't sure if Product Model is a new term alongside Resource Type, whereas it seems like we only need the latter in combination with the other existing terms..

**Comment by bhaugen** [comment]
ProductModel is [schema.org's term](https://schema.org/ProductModel) for "A datasheet or vendor specification of a product (in the sense of a prototypical description)." Which is pretty much like a subset of possible ResourceTypes.

I think @elf-pavlik cited it as an example of existing vocab terms that we could reference in one way or another. But he can explain more if and when he wants. As @fosterlynn wrote above, it's not equivalent to rea: or vf:ResourceType.

But he is correct that people who want to coordinate will need to agree on ResourceTypes.

**Comment by elf-pavlik** [comment]
> But he is correct that people who want to coordinate will need to agree on ResourceTypes.

@ahdinosaur I would still like to know how you plan to relate *ex:apple* product model, specific to particular agent where we have as many product models as typsnowflakes, with some common reference for resource types like *prodont:Apple*

**Comment by ahdinosaur** [comment]
> I would still like to know how you plan to relate ex:apple product model, specific to particular agent where we have as many product models as types of snowflakes, with some common reference for resource types like prodont:Apple or agrovoc:apples

https://github.com/valueflows/process/issues/3#issuecomment-219600972

**Comment by elf-pavlik** [comment]
SKOS relates concepts (instances of skos:Concept) to each other, so far we looked at defining types as instances of rdfs:Class. 
I already referenced this document:
* https://www.w3.org/2006/07/SWD/SKOS/skos-and-owl/master.html

Those two also come relevant:
* https://www.w3.org/TR/skos-reference/#L1170
* https://www.w3.org/TR/2005/WD-swbp-skos-core-guide-20051102/#secmodellingrdf

**Comment by fosterlynn** [comment]
>And I think we will need to do all of that. I don't know much we need to put into the VF vocab and how much to re-use. Certainly the standard conversions can come from qudt.

Seems like we just need a unit in the vf:Resource so that people know what the quantity is about.  Then conversion can happen whenever / however.  If we eventually want to support knowing more than one equivalent quantity/unit for one vf:Resource, then that is easy to add on I think.  I'm not sure you would even need to change the vocab, you can always define multiples of any property.  
( @elf-pavlik am I correct on this, or did I just make this up?)

**Comment by bhaugen** [comment]
If a resource has a quantity+unit (number + unit of measure) and they always go together (as I understand they do in qudt) then a resource could have multiples of those and unit conversions should still work. As long as the multiple qty+units are consistent with each other. That is, if I understand my qudts correctly, they should be the same Quantity Kind (like Length or Mass, etc, but not some Length units and some Mass units) and their numeric values should be accurate conversions from one unit to another.

**Comment by elf-pavlik** [comment]
*vf:amount*, as any other propertynot define as *owl:FunctionalProperty*, can have multiple values, each of them an instance of *qudt:QuantityValue*. Than each of those instances can have different unit!

**Comment by elf-pavlik** [comment]
I added example which uses `@type` for:
* VF class
* the generic type (from product ontology)
* producer specific model
* https://github.com/valueflows/resource/blob/master/Resource.md#bicycle

**Comment by elf-pavlik** [comment]
Possibly also worth keeping in mind https://en.wikipedia.org/wiki/Abox
* generic taxonomies seem somewhere in between Tbox and Abox
* producer specific models seem to fit Abox

**Comment by elf-pavlik** [comment]
from https://docs.google.com/drawings/d/13NxkdreIQ_0H43H25kBC5ZKYz2QCgfk7XalOXExUK7Q/edit

> @elf-pavlik 
resourceClass property? [on nrp:ResourceType] values of what type would appear here?

> @fosterlynn 
We added that to say something like "carrot" so that all resource types that were carrots but had some transformation through workflow processes could know they were the same class of thing. Haven't used it much. For a while in DHEN we had Catnip Fresh and Catnip Dry as 2 resource types, that both were class Catnip.

> @elf-pavlik
hat again sounds relevant to 'Product Model' and 'Resource Type' discussion...

> @fosterlynn 
Could be. I'm not advocating we have resource class in the vocab. And this picture doesn't do any subclassing, so there will be elements that only apply to a subclass. The point is to compare the layers, and this is something people can do on the type layer. Probably lots more, ala ProductModel

**Comment by elf-pavlik** [comment]
Currently I think in this direction:

* reserve `@type` (`rdf:type`) for VF class, like *vf:Resource* or *vf:Service*
* introduce *vf:kind* of *vf:category* for generic, common resource types from sources like wikidata, product ontology (wikipedia based), agrovoc. working on converging to use same reference globally
* introduce *vf:model* for vendor specific product models, so john's potatoes and marry's potatoes will have same *vf:kind* (or *vf:category*) but distinct *vf:model*

**Comment by fosterlynn** [comment]
@elf-pavlik yes, agree!!  (may need to work on the names a bit, but the conceptual split is very helpful!)

**Comment by fosterlynn** [comment]
I think I am liking vf:category better than vf:kind.  We can use the same thing for categorizing intents then.  I'm trying it out in yaml now.

**Comment by elf-pavlik** [comment]
https://en.wikipedia.org/wiki/Stock_keeping_unit

> In the field of inventory management, a stock keeping unit (SKU) (/ˌɛsˌkeɪˈjuː/, /ˈskuː/ or /ˈskjuː/) is a distinct type of item for sale,[1] such as a product or service, and all attributes associated with the item type that distinguish it from other item types. For a product, these attributes could include, but are not limited to, manufacturer, description, material, size, color, packaging, and warranty terms. When a business takes an inventory, it counts the quantity it has of each SKU.

> SKU can also refer to a unique identifier or code that refers to the particular stock keeping unit. These codes are not regulated or standardized. When a company receives items from a vendor, it has a choice of maintaining the vendor's SKU or creating its own. Other entity tracking methods, with varying regulations, are the Universal Product Code (UPC), International Article Number (EAN), Global Trade Item Number (GTIN), and Australian Product Number (APN).

I think in some cases we seem to need just SKU, besides references to common Taxonomies or common open source designs.

**Comment by elf-pavlik** [comment]
continuing from https://github.com/valueflows/valueflows/issues/300#issuecomment-442450129
> But in case of lemon pie, lemon input https://www.wikidata.org/wiki/Q1093742 in most cases seems more than enough. (in *consume* recipe flow)

**Comment by fosterlynn** [comment]
@elf-pavlik how about this?  Or not include EconomicEvent, just assume in that case it is all in EconomicResource?  Or doesn't make sense at all...?

```
vf:resourceConformsTo
        a               owl:ObjectProperty ;
        rdfs:label      "conforms to" ;
        rdfs:domain     [ owl:unionOf (vf:Commitment vf:EconomicEvent vf:EconomicResource vf:Intent vf:RecipeFlow ) ] ;
        rdfs:range      vf:ResourceSpecification ;

vf:resourceClassifiedAs
        a                   owl:ObjectProperty ;
        rdfs:label          "classified as" ;
        rdfs:domain       [ owl:unionOf (vf:Commitment vf:EconomicEvent vf:EconomicResource vf:Intent vf:ResourceSpecification ) ] ;
        rdfs:range          owl:Thing ;
```
